Python upcoming events is done.

behumb opened the pull request.

behumb committed:
Python upcoming events is done.

denisorehovsky reviewed (changes requested):
You need to get start date and end date of an event. I think you need to scrape every event page. Not only home page

behumb committed:
Algorithm has been change. Now program scrape not home events page but each page of event

denisorehovsky commented on python_upcoming-events/main.py:
```diff
@@ -0,0 +1,45 @@
+if __name__ == '__main__':
+    python_main_link = 'https://www.python.org'
+    soup = get_soup(python_main_link)
+    events_btn = soup.find('li', attrs={'id': 'events'}).a
```
It's difficult to read this code. Why do you have a separate function for getting events info, but don't have a function for getting event urls?

if __name__ == '__main__' should be as small as possible

denisorehovsky commented on python_upcoming-events/main.py:
```diff
@@ -0,0 +1,45 @@
+    events_btn = soup.find('li', attrs={'id': 'events'}).a
+
+    python_events_link = python_main_link + events_btn['href']
+    soup = get_soup(python_events_link)
```
Why do you need to call get_soup multiple times? You only need to make one http request in order to get a list of event urls, no?

denisorehovsky commented on python_upcoming-events/main.py:
```diff
@@ -0,0 +1,45 @@
+    reformat_date = dt.strptime(date, '%Y-%m-%dT%H:%M:%S%z')
+    return reformat_date.strftime('%d %B %Y')
+
+def get_events_info(links):
```
I think it would be better if you also had `get_event_info` function. This function takes one parameter (link) and returns information about the event.

denisorehovsky replied:
Also, I would consider using python threads here. Try to investigate whether the performance increases or not if you send a request to the event detail page in a separate thread.

denisorehovsky commented on python_upcoming-events/main.py:
```diff
@@ -0,0 +1,45 @@
+    events_soups = [get_soup(link) for link in links]
+    events_info = []
+    for soup in events_soups:
+        events_info.append({'title': soup.select_one('h1.single-event-title').text.strip(),
```
Instead of creating a dictionary for storing information about events, I would create a named tuple instead. It fits perfectly here.

denisorehovsky reviewed (changes requested):
Also, read about commit messages. For example, this one https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

behumb committed:
Rework function of getting events links

 - Change usual request to site to request with help of threads.
It allowed to reduce the time of access and data acqusition.

 - Fix problem with overflowing of "if '__name__' == '__main'".

 - Change storage of events info. Dict was replaced by namedtuple.
 It allowed to make code more readable and convenient.

behumb committed:
Rework main algorithm of accessing events info

 - Rework usual site data query to query using threads. It allowed
 to reduce accessing time.
 - Fix ploblem with overflowing of "if 'name' == 'main'". Logic was
 moved to get_events_links function. Code has become more readable.
 - Change a storage of events info. Dict was replaced to namedtuple.

denisorehovsky commented on python_upcoming-events/main.py:
```diff
@@ -0,0 +1,55 @@
+def get_events_info(main_link):
+    events_links = get_events_links(main_link)
+    events_info = []
+    EventInfo = namedtuple('EventInfo', 'title location date_start date_end')
```
Better to name it `Event` and also create it outside of the function.

denisorehovsky commented on python_upcoming-events/main.py:
```diff
@@ -0,0 +1,55 @@
+            soup.select_one('span.single-event-location').text.strip(),
+            reformat_event_date(soup.select_one('time.date-start')['datetime']),
+            reformat_event_date(soup.select_one('time.date-end')['datetime'])]
+
```
Read about PEP8. You should have two spaces between functions.

denisorehovsky replied:
https://pep8.org/

denisorehovsky commented on python_upcoming-events/main.py:
```diff
@@ -0,0 +1,55 @@
+
+def get_event_info(link):
+    soup = get_soup(link)
+    return [soup.select_one('h1.single-event-title').text.strip(),
```
Why do you need to create a list and then create EventInfo from this list? I think this function should create and return EventInfo. There is no need to create a list.

denisorehovsky reviewed (changes requested):
Looks good I think. Some small things left.

behumb committed:
Fix some small issues

denisorehovsky commented on python_upcoming-events/main.py:
```diff
@@ -0,0 +1,68 @@
+def reformat_event_date(date):
+    reformat_date = dt.strptime(date, '%Y-%m-%dT%H:%M:%S%z')
+    return reformat_date
+
```
Should be two blank lines

denisorehovsky commented on python_upcoming-events/main.py:
```diff
@@ -0,0 +1,68 @@
+    soup = get_soup(link)
+
+    return Event(soup.select_one('h1.single-event-title').text.strip(),
+                    soup.select_one('span.single-event-location').text.strip(),
```
This line is not formatted correctly as well

denisorehovsky commented on python_upcoming-events/main.py:
```diff
@@ -0,0 +1,68 @@
+
+        for future in concurrent.futures.as_completed(futures):
+            events_info.append(future.result())
+    events_info.sort(key=attrgetter('date_start'))
```
I think it's not the best solution. Did you try to use map instead of submit? https://stackoverflow.com/questions/20838162/how-does-threadpoolexecutor-map-differ-from-threadpoolexecutor-submit#:~:text=They%20both%20accept%20the%20jobs,the%20order%20they%20are%20submitted.

behumb committed:
Update method of processing threads

behumb committed:
Delete sorting of events by start date

denisorehovsky reviewed (approved):
Perfect
